### ISSUE TITLE
Preserving non-numeric (or non-positive-integer) array keys

### ISSUE THREAD

**Issue by brettz9** [dfahlander/typeson#9]
In IndexedDBShim, I ran across some W3C test cases where non-numeric properties were injected onto arrays, and due to Typeson not handling such currently (as array properties won't survive the `JSON.stringify` calls used internally), they failed. Likewise, I've noticed that the Structured Cloning Algorithm in general calls for the preservation of non-numeric/non-positive-integer keys on arrays as with objects which also needs support in IndexedDBShim.

I tried my hand at creating a type using the `testPlainObjects` config which could find arrays and convert them to objects (using `Object.fromEntries(Object.entries(...))` which preserves non-positive-integer properties and avoids explicit `undefined` for sparsed arrays), but approaching this through creating a type appears it would be very difficult to get to work with cyclics. (I had thought of something similar for the proposal for issue #6 but here too I am not sure how to make things work with cyclic tracking when types themselves may have cyclics...)

I am therefore considering adding a mode directly to Typeson to temporarily convert arrays to objects, while allowing for cyclics.

Since you came up with the cyclic code, I didn't know if you had any thoughts about this, David @dfahlander . Thanks!

**Comment by brettz9** [comment]
Oh, you might want to also take a look at a bunch of refactoring I've done--I was finding it hard to keep track of things as the code expanded, so switched to ES6 class in source, added JSDoc, etc.

**Comment by dfahlander** [comment]
Thanks @brettz9! I think it's great you've converted to ES6. I was using typeson recently in a new dexie addon [dexie-export-import](https://github.com/dfahlander/Dexie.js/tree/master/addons/dexie-export-import) and found some issues that I could work around locally, but I was going to contact you about it. Time was too short at the time, so I'm glad you bring it up.

I stumbled on the issue with converting binaries to UTF-16 was not working correctly for all values. In my case, I make an alternate type handler for ArrayBuffer that converted to base64. Also had issues with that I was unable to npm install typeson-registry on my mac due to native dependencies. When I did that on my PC the other year, that depencency could be resolved by installing it manually but that seems not to be possible on mac. Would it be possible to make that dependency optional and skip unit tests on Image if it's not installed? I also never got the sync version of Blob reading to work. Tested with a blob containing all bytes from 0..255 and some numbers failed to be read correctly. So I had to use encapsulateAsync() witch did not work either, so I ended up making another sync solution. See [tson.ts](https://github.com/dfahlander/Dexie.js/blob/master/addons/dexie-export-import/src/tson.ts) in the addon source.

I also understand the issue with sparse arrays and object-like arrays. Don't know why the cyclic check fails though and would need time to investigate it, but right now I am in the middle of a customer project with a harsh deadline so this would not be the time. Hoping to get more time for typeson further on.

**Comment by brettz9** [comment]
I really think both sync and async encapsulation/stringification (and revival) should be fixed now for Blob/File with recent commits 6c9c650b624e79a2895cdc98eafdf799c4bc6c70 and fe0acb57ba98cb573dc71e12f18a4316450f9535 (alpha 22 and alpha 23). I added more to the large value test and more IndexedDBShim tests are passing now too.

(Note though that I am relying on synchronous XHR for the sync method, so if that is ever removed--as per the frustrating deprecation showing now in browser consoles--users would need to move to async or we'd need to incorporate your approach for workers-only.)

Regarding the dependency, I guess you mean `canvas`, and sure that could be avoided, but that is only a devDependency... Were you trying a local install?  Also, canvas is working for me on my Mac (Mojave), using nvm (currently with Node 10.11.0 though).

**Comment by brettz9** [comment]
Regarding the stumbling block regarding cyclics and preserving non-numeric array properties, no worries then, but I'll spell out the issue in some more detail for possible future reference (in case I don't come up with a satisfactory solution before then) or maybe just for my own clarity.

I realized we already have a mechanism (through `testPlainObjects` and `stateObj.iterateIn`) we can use for storing arrays as objects and thus preserving their non-numeric/non-pos.-integer keys (without using `Object.fromEntries(Object.entries(...))`) (though I needed to make a fix a couple days ago to get it working as intended). The second example below should make clear how this works.

However, the issue still remains of how to determine precedence vis-a-vis reviving types vs. resolving cyclics (including shared references).

Here is a case where we have an object with a type (`DataView`) containing a reference shared with another property (an `ArrayBuffer`) and whose serialized buffer "cyclic" reference must be resolved and set *before* revival (i.e., the `DataView` cannot be expected to revive a buffer defined simply as "#buf"). Our code currently works with this example.


```js

    const buf = new ArrayBuffer(16);
    const data = {
        buf,
        bar: {
            data: new DataView(buf, 8, 8)
        }
    };

    const tson = typeson.stringify(data, null, 2);

```

```json

{
  "buf": "AAAAAAAAAAAAAAAAAAAAAA==",
  "bar": {
    "data": {
      "buffer": "#buf",
      "byteOffset": 8,
      "byteLength": 8
    }
  },
  "$types": {
    "buf": "ArrayBuffer",
    "bar.data": "DataView",
    "bar.data.buffer": "#"
  }
}

```


However, here is a case, which our code doesn't work with currently, where the references must be resolved and set *after* type revival (since otherwise, some references will be left referring to the encapsulated object rather than pointing to the array which the object gets revived as):


```js

        const typeson = new Typeson().register({
            arraysToObjects: {
                testPlainObjects: true,
                test (x, stateObj) {
                    if (Array.isArray(x)) {
                        stateObj.iterateIn = 'object';
                        stateObj.addLength = true;
                        return true;
                    }
                    return false;
                },
                revive (o) {
                    const arr = new Array(o.length);
                    Object.entries(o).forEach(([key, val]) => {
                        if (key === 'length') {
                            return;
                        }
                        arr[key] = val;
                    });
                    return arr;
                }
            }
        });
        const arr = new Array(10);
        arr[0] = 3;
        arr[3] = '4';
        arr[4] = 5;
        arr[6] = arr;
        arr[9] = {arr};
        arr.b = 'ddd';
        arr[-2] = 'eee';

        const tson = typeson.stringify(arr);

```


```json
{
  "0": 3,
  "3": "4",
  "4": 5,
  "6": "#",
  "9": {
    "arr": "#"
  },
  "length": 10,
  "b": "ddd",
  "-2": "eee",
  "$types": {
    "6": "#",
    "": "arraysToObjects",
    "9.arr": "#"
  }
}
```


Maybe this is just a case of only `testPlainObjects` references needing to be delayed, but I haven't thought it through to see whether other cases may apply. I'd also think there may be cases where one needs `testPlainObjects` references resolved and set first so as to be used within nested types (such as if the first example depended on an array internally instead of a buffer), though the references would need to be revived (e.g., as arrays) first.

**Comment by brettz9** [comment]
I've gotten something that needs a little more testing, but it looks like reviving `testPlainObjects` types first is doing the trick...

**Comment by brettz9** [comment]
This is now fixed in `5.11.0`/`master`
